### PR TITLE
Feature/slow down dynamic obs

### PR DESCRIPTION
**Pull request by grieneis**

### CODE REVIEW

File: base_local_planner/include/base_local_planner/critics/speed_cost_function.h [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    calculateStopDistances();
+  }
+
+  void setWorldFrameId(std::string id)
+  {
+    world_frame_id_ = id;
+  }
+
+  // void setCircumscribedRadius(double rad)
+  // {
```
Comment (dantwinkler) [inline base_local_planner/include/base_local_planner/critics/speed_cost_function.h]: Remove?

File: base_local_planner/include/base_local_planner/critics/speed_cost_function.h [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+  /**
+   * Scores the trajectory.  Returns a negative value for rejected trajectories.
+   * @param traj The trajectory
+   * @return Non-negative value if the trajectory is valid, negative otherwise.
+   *     -1 indicates that it is in violation of the distance to obstruction metric
+   */
+  double scoreTrajectory(Trajectory &traj);
+
+  void setMaxVelocity(double vel) {
```
Comment (dantwinkler) [inline base_local_planner/include/base_local_planner/critics/speed_cost_function.h]: Name should match variable (linear).

File: base_local_planner/src/critics/speed_cost_function.cpp [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+      continue;
+    }
+
+    double speed = calculateAllowedSpeed(obs_body_frame);
+    if (speed < max_allowed_vel_)
+    {
+      max_allowed_vel_ = speed;
+    }
+  }
+  ROS_INFO_THROTTLE(0.2, "Setting max speed to %f", max_allowed_vel_);
```
Comment (dantwinkler) [inline base_local_planner/src/critics/speed_cost_function.cpp]: Should probably move this into the if that sets the speed, not sure how spammy this is but maybe should be a DEBUG?

File: base_local_planner/src/critics/speed_cost_function.cpp [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+costmap_2d::ObstructionMsg SpeedCostFunction::obstructionToBodyFrame(const costmap_2d::ObstructionMsg& in)
+{
+  if (in.frame_id == body_frame_id_)
+  {
+    return in;
+  }
+  else if (in.frame_id != world_frame_id_)
+  {
+    ROS_WARN_THROTTLE(1.0, "Received obstruction with unknown frame_id %s", in.frame_id.c_str());
```
Comment (dantwinkler) [inline base_local_planner/src/critics/speed_cost_function.cpp]: Should this be an ERROR?

File: base_local_planner/src/critics/speed_cost_function.cpp [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+void SpeedCostFunction::calculateStopDistances()
+{
+  max_distance_to_stop_ = 0.5 * max_linear_velocity_ * max_linear_velocity_ / linear_acceleration_;
+  min_distance_to_stop_ = 0.5 * min_linear_velocity_ * min_linear_velocity_ / linear_acceleration_;
+}
+
+void SpeedCostFunction::calculateFootprintBounds(std::vector<geometry_msgs::Point> footprint)
+{
+  if (footprint.empty())
```
Comment (dantwinkler) [inline base_local_planner/src/critics/speed_cost_function.cpp:184]: Is this a valid case or an ERROR?

**Review by dantwinkler — APPROVED** [review]
Looks good, just a few comments.

File: base_local_planner/include/base_local_planner/critics/speed_cost_function.h (hunk 1 above)
Comment (grieneis) [reply base_local_planner/include/base_local_planner/critics/speed_cost_function.h]: Done

File: base_local_planner/include/base_local_planner/critics/speed_cost_function.h (hunk 2 above)
Comment (grieneis) [reply base_local_planner/include/base_local_planner/critics/speed_cost_function.h]: Done

File: base_local_planner/src/critics/speed_cost_function.cpp (hunk 3 above)
Comment (grieneis) [reply base_local_planner/src/critics/speed_cost_function.cpp]: Dpme

File: base_local_planner/src/critics/speed_cost_function.cpp (hunk 4 above)
Comment (grieneis) [reply base_local_planner/src/critics/speed_cost_function.cpp]: Done

File: base_local_planner/src/critics/speed_cost_function.cpp (hunk 5 above)
Comment (grieneis) [reply base_local_planner/src/critics/speed_cost_function.cpp:184]: Error.  Added an error message.

**Pull request merged**